external/urlfrontier/src/main/java/com/digitalpebble/stormcrawler/urlfrontier/StatusUpdaterBolt.java

```diff
@@ -170,20 +171,28 @@ public void prepare(
         }
 
         if (address == null) {
-            channel =
-                    ManagedChannelUtil.createChannel(
-                            ConfUtils.getString(
-                                    stormConf, URLFRONTIER_HOST_KEY, URLFRONTIER_DEFAULT_HOST),
-                            ConfUtils.getInt(
-                                    stormConf, URLFRONTIER_PORT_KEY, URLFRONTIER_DEFAULT_PORT));
-        } else {
-            channel = ManagedChannelUtil.createChannel(address);
+            String host =
+                    ConfUtils.getString(stormConf, URLFRONTIER_HOST_KEY, URLFRONTIER_DEFAULT_HOST);
+            int port = ConfUtils.getInt(stormConf, URLFRONTIER_PORT_KEY, URLFRONTIER_DEFAULT_PORT);
+            address = host + ":" + port;
         }
 
+        channel = ManagedChannelUtil.createChannel(address);
+        channel.notifyWhenStateChanged(ConnectivityState.SHUTDOWN, () -> onChannelStateChange(ConnectivityState.SHUTDOWN));
```

Comment:
this is an elegant way of dealing with it. Have you tried this on a crawl?

Reply:
I did one test crawl, in which I was in particular focusing on whether the crawler is able to continue as usual, after stopping the frontier for some time and restarting it then again. This work for me. Within the real crawling setup, the `notifyWhenStateChanged` callback is necessary and the restarting wouldn't have worked without using the callback. Remarkably, not so in the simpler test environment.